### PR TITLE
Fix computation of the main module name

### PR DESCRIPTION
**Pull request by ghost**
The previous code was failing too early when the library wasn't available, for instance when it was optional with missing dependencies.

### CODE REVIEW

**Review by emillon — COMMENTED** [review]
Do you think it's worth adding a regression test for this issue?

File: src/dune_file.ml [hunk 1]
```diff
@@ -1065,12 +1065,18 @@ module Library = struct
 
   let is_virtual t = Option.is_some t.virtual_modules
 
-  let main_module_name t =
+  module Main_module_name = struct
+    type t =
+      | This of Module.Name.t option
+      | Inherited_from of (Loc.t * Lib_name.t)
```
Comment (rgrinberg) [inline src/dune_file.ml]: Looks good. I actually had a wip refactoring that introduced a type like:

```
module Module : sig
  module Name : sig
    module Main : sig
      type t = Named of Name.t | Inherited_from of (Loc.t * Lib_name.t)
    end
  ..
  end
  ..
end
```

and the return value for all these function was `Module.Name.Main.t option`. But I think that this basically equivalent. Maybe we should move this sub module to `module.ml` however.

**Review by rgrinberg — APPROVED** [review]
Thanks for the nice fix 💯  Sorry for this bug

**Comment by rgrinberg** [comment]
> Do you think it's worth adding a regression test for this issue?

Note that various forms of this issue have appeared and it seems like it's a more fundamental issue. That being said, if it's not too much trouble, a regression test would be nice.

**Comment by diml** [comment]
Sure, I added a regression test

**Comment by diml** [comment]
The first commit exposes the issue and we can see that the second fixes it

**Review by emillon — APPROVED** [review]

**Pull request merged**